File: tts-api/sdk/javascript/tts_client.js
/**
 * sdk/javascript/tts_client.js
 *
 * Lightweight JavaScript / Node.js client for the TTS AI API.
 */
const fs = require('fs');
const path = require('path');

class TTSClient {
    /**
     * @param {Object} config
     * @param {string} [config.baseUrl='http://localhost:8000']
     * @param {string} [config.apiKey=null]
     */
    constructor({ baseUrl = 'http://localhost:8000', apiKey = null } = {}) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.apiKey = apiKey;
    }

    _getHeaders(extra = {}) {
        const headers = {
            'Content-Type': 'application/json',
            ...extra,
        };
        if (this.apiKey) {
            headers['X-API-Key'] = this.apiKey;
        }
        return headers;
    }

    /**
     * Check API health status.
     */
    async checkHealth() {
        const res = await fetch(`${this.baseUrl}/api/v1/health`, {
            headers: this._getHeaders(),
        });
        if (!res.ok) throw new Error(`Health check failed: ${res.statusText}`);
        return await res.json();
    }

    /**
     * List available voices.
     */
    async listVoices({ language = null, gender = null } = {}) {
        const url = new URL(`${this.baseUrl}/api/v1/voices`);
        if (language) url.searchParams.append('language', language);
        if (gender) url.searchParams.append('gender', gender);

        const res = await fetch(url.toString(), {
            headers: this._getHeaders(),
        });
        if (!res.ok) throw new Error(`List voices failed: ${res.statusText}`);
        const data = await res.json();
        return data.voices || [];
    }

    /**
     * Synthesize speech synchronously and return an ArrayBuffer of audio bytes.
     */
    async synthesize({ text, voice = 'af_sarah', language = 'en', speed = 1.0, format = 'wav' }) {
        const res = await fetch(`${this.baseUrl}/api/v1/tts`, {
            method: 'POST',
            headers: this._getHeaders({ Accept: `audio/${format}` }),
            body: JSON.stringify({
                text,
                voice,
                language,
                speed,
                format,
                return_json: false,
            }),
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(`TTS failed (${res.status}): ${err}`);
        }
        return await res.arrayBuffer();
    }

    /**
     * Synthesize speech and save to local disk file (Node.js).
     */
    async synthesizeToFile({ text, outputPath, voice = 'af_sarah', language = 'en', speed = 1.0, format = 'wav' }) {
        const buffer = await this.synthesize({ text, voice, language, speed, format });
        const dir = path.dirname(outputPath);
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir, { recursive: true });
        }
        fs.writeFileSync(outputPath, Buffer.from(buffer));
        return outputPath;
    }

    /**
     * Submit an asynchronous TTS job for long-form content.
     */
    async synthesizeAsync({ text, voice = 'af_sarah', language = 'en', speed = 1.0, format = 'wav', webhookUrl = null }) {
        const res = await fetch(`${this.baseUrl}/api/v1/tts/async`, {
            method: 'POST',
            headers: this._getHeaders(),
            body: JSON.stringify({
                text,
                voice,
                language,
                speed,
                format,
                webhook_url: webhookUrl,
            }),
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(`Async TTS submission failed: ${err}`);
        }
        const data = await res.json();
        return data.job_id;
    }

    /**
     * Poll an async job until completion.
     */
    async pollJob(jobId, { intervalMs = 1000, timeoutMs = 120000 } = {}) {
        const startTime = Date.now();
        while (Date.now() - startTime < timeoutMs) {
            const res = await fetch(`${this.baseUrl}/api/v1/jobs/${jobId}`, {
                headers: this._getHeaders(),
            });
            if (!res.ok) throw new Error(`Job poll failed: ${res.statusText}`);
            const data = await res.json();
            if (data.status === 'completed') {
                return data;
            } else if (data.status === 'failed' || data.status === 'cancelled') {
                throw new Error(`Job ${jobId} failed: ${data.error_message || data.status}`);
            }
            await new Promise((r) => setTimeout(r, intervalMs));
        }
        throw new Error(`Job ${jobId} timed out after ${timeoutMs}ms.`);
    }
}

module.exports = { TTSClient };

File: tts-api/sdk/python/tts_client.py
"""
sdk/python/tts_client.py

Lightweight Python Client SDK for consuming the TTS AI API.

Usage:
    from sdk.python.tts_client import TTSClient

    client = TTSClient(base_url="http://localhost:8000", api_key="YOUR_API_KEY")
    client.synthesize_to_file(
        text="Hello from my Python application!",
        voice="af_sarah",
        language="en",
        output_path="speech.wav"
    )
"""
from __future__ import annotations

import time
from pathlib import Path
from typing import Any, Dict, List, Optional

import httpx


class TTSError(Exception):
    """Exception raised by TTSClient operations."""
    pass


class TTSClient:
    """
    Official Python SDK for the Production TTS AI Platform.
    """

    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        api_key: Optional[str] = None,
        timeout: float = 60.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self._headers: Dict[str, str] = {
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
        if self.api_key:
            self._headers["X-API-Key"] = self.api_key

    def check_health(self) -> Dict[str, Any]:
        """Check API service health status."""
        url = f"{self.base_url}/api/v1/health"
        with httpx.Client(timeout=self.timeout) as client:
            resp = client.get(url, headers=self._headers)
            if resp.status_code != 200:
                raise TTSError(f"Health check failed with status {resp.status_code}: {resp.text}")
            return resp.json()

    def list_voices(self, language: Optional[str] = None, gender: Optional[str] = None) -> List[Dict[str, Any]]:
        """List available TTS voices."""
        url = f"{self.base_url}/api/v1/voices"
        params = {}
        if language:
            params["language"] = language
        if gender:
            params["gender"] = gender

        with httpx.Client(timeout=self.timeout) as client:
            resp = client.get(url, headers=self._headers, params=params)
            if resp.status_code != 200:
                raise TTSError(f"Listing voices failed: {resp.text}")
            return resp.json().get("voices", [])

    def synthesize(
        self,
        text: str,
        voice: str = "af_sarah",
        language: str = "en",
        speed: float = 1.0,
        format_ext: str = "wav",
    ) -> bytes:
        """
        Synthesize speech synchronously and return audio bytes directly.
        """
        url = f"{self.base_url}/api/v1/tts"
        payload = {
            "text": text,
            "voice": voice,
            "language": language,
            "speed": speed,
            "format": format_ext,
            "return_json": False,
        }
        headers = {**self._headers, "Accept": f"audio/{format_ext}"}

        with httpx.Client(timeout=self.timeout) as client:
            resp = client.post(url, json=payload, headers=headers)
            if resp.status_code != 200:
                raise TTSError(f"Synthesis failed with HTTP {resp.status_code}: {resp.text}")
            return resp.content

    def synthesize_to_file(
        self,
        text: str,
        output_path: str | Path,
        voice: str = "af_sarah",
        language: str = "en",
        speed: float = 1.0,
        format_ext: str = "wav",
    ) -> Path:
        """
        Synthesize speech and write audio to a local file.
        """
        out_path = Path(output_path)
        out_path.parent.mkdir(parents=True, exist_ok=True)
        audio_bytes = self.synthesize(
            text=text,
            voice=voice,
            language=language,
            speed=speed,
            format_ext=format_ext,
        )
        with open(out_path, "wb") as f:
            f.write(audio_bytes)
        return out_path

    def synthesize_async(
        self,
        text: str,
        voice: str = "af_sarah",
        language: str = "en",
        speed: float = 1.0,
        format_ext: str = "wav",
        webhook_url: Optional[str] = None,
    ) -> str:
        """
        Submit a long-form TTS job for background synthesis. Returns job_id.
        """
        url = f"{self.base_url}/api/v1/tts/async"
        payload = {
            "text": text,
            "voice": voice,
            "language": language,
            "speed": speed,
            "format": format_ext,
            "webhook_url": webhook_url,
        }

        with httpx.Client(timeout=self.timeout) as client:
            resp = client.post(url, json=payload, headers=self._headers)
            if resp.status_code not in (200, 202):
                raise TTSError(f"Async synthesis submission failed: {resp.text}")
            return resp.json()["job_id"]

    def poll_job(self, job_id: str, poll_interval: float = 1.0, max_wait: float = 120.0) -> Dict[str, Any]:
        """
        Poll job status until completion or failure.
        """
        url = f"{self.base_url}/api/v1/jobs/{job_id}"
        start_time = time.time()

        with httpx.Client(timeout=self.timeout) as client:
            while time.time() - start_time < max_wait:
                resp = client.get(url, headers=self._headers)
                if resp.status_code != 200:
                    raise TTSError(f"Polling job failed: {resp.text}")
                data = resp.json()
                if data["status"] == "completed":
                    return data
                elif data["status"] in ("failed", "cancelled"):
                    raise TTSError(f"Job {job_id} terminated with status: {data['status']} - Error: {data.get('error_message')}")
                time.sleep(poll_interval)

        raise TTSError(f"Job {job_id} timed out after {max_wait} seconds.")

    def download_audio(self, audio_id: str, output_path: str | Path, format_ext: str = "wav") -> Path:
        """
        Download previously synthesized audio file by audio_id.
        """
        url = f"{self.base_url}/api/v1/audio/{audio_id}?format={format_ext}&download=true"
        out_path = Path(output_path)
        out_path.parent.mkdir(parents=True, exist_ok=True)

        with httpx.Client(timeout=self.timeout) as client:
            resp = client.get(url, headers=self._headers)
            if resp.status_code != 200:
                raise TTSError(f"Downloading audio failed: {resp.text}")
            with open(out_path, "wb") as f:
                f.write(resp.content)
        return out_path
